blockchain/chain_test.go

```diff
@@ -1311,3 +1313,381 @@ func TestIsAncestor(t *testing.T) {
+			},
+		},
+		{
+			name: "invalidate a side branch",
+			chainGen: func() (*BlockChain, []*chainhash.Hash, func()) {
+				source := rand.NewSource(0)
+				rand := rand.New(source)
+
+				chain, params, tearDown := utxoCacheTestChain("TestInvalidateBlock-invalidate-side-branch")
+				tip := btcutil.NewBlock(params.GenesisBlock)
+
+				// Create block at height 1.
+				var emptySpendableOuts []*testhelper.SpendableOut
+				b1, spendableOuts1, err := addBlock(chain, tip, emptySpendableOuts)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				var allSpends []*testhelper.SpendableOut
+				nextBlock := b1
+				nextSpends := spendableOuts1
+
+				// Create a chain with 11 blocks.
+				for b := 0; b < 10; b++ {
+					newBlock, newSpendableOuts, err := addBlock(chain, nextBlock, nextSpends)
+					if err != nil {
+						t.Fatal(err)
+					}
+
+					nextBlock = newBlock
+
+					allSpends = append(allSpends, newSpendableOuts...)
+
+					var nextSpendsTmp []*testhelper.SpendableOut
+					for i := 0; i < len(allSpends); i++ {
+						randIdx := rand.Intn(len(allSpends))
+
+						spend := allSpends[randIdx]                                       // get
+						allSpends = append(allSpends[:randIdx], allSpends[randIdx+1:]...) // delete
+						nextSpendsTmp = append(nextSpendsTmp, spend)
+					}
+					nextSpends = nextSpendsTmp
+
+					if b%10 == 0 {
+						// Commit the two base blocks to DB
+						if err := chain.FlushUtxoCache(FlushRequired); err != nil {
+							t.Fatalf("unexpected error while flushing cache: %v", err)
+						}
+					}
+				}
+
+				// Create a side chain with 7 blocks that builds on block 1.
+				var altSpends []*testhelper.SpendableOut
+				altNextSpends := spendableOuts1
+				altNextBlock := b1
+				var invalidateHash *chainhash.Hash
+				for b := 0; b < 6; b++ {
+					altNewBlock, newSpends, err := addBlock(chain, altNextBlock, altNextSpends)
+					if err != nil {
+						t.Fatal(err)
+					}
+
+					altNextBlock = altNewBlock
+
+					altSpends = append(altSpends, newSpends...)
+
+					if altNewBlock.Height() == 4 {
+						invalidateHash = altNewBlock.Hash()
+					}
+
+					var nextSpendsTmp []*testhelper.SpendableOut
+					for i := 0; i < len(altSpends); i++ {
+						randIdx := rand.Intn(len(altSpends))
+
+						spend := altSpends[randIdx]                                       // get
+						altSpends = append(altSpends[:randIdx], altSpends[randIdx+1:]...) // delete
+						nextSpendsTmp = append(nextSpendsTmp, spend)
+					}
+					altNextSpends = nextSpendsTmp
+
+					if b%10 == 0 {
+						// Commit the two base blocks to DB
+						if err := chain.FlushUtxoCache(FlushRequired); err != nil {
+							t.Fatalf("unexpected error while flushing cache: %v", err)
+						}
+					}
+				}
+
+				return chain, []*chainhash.Hash{invalidateHash}, tearDown
+			},
+		},
+	}
+
+	for _, test := range tests {
+		chain, invalidateHashes, tearDown := test.chainGen()
+		func() {
```

Comment:
Why the closure func that gets immediately executed?

Reply:
We can also use `t.Run` here, then that way the test name gets printed during execution, and we can run just one of the sub-tests by name.

Reply:
Was doing this so that even when the test fails, we still run the teardown